http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandlerAlgebra.scala

```diff
@@ -168,19 +167,31 @@ object AzurePubsubHandler {
         .take(size)
         .toList
 
-    val passwordChars: List[Char] = pickRandomChars(lowerLetters, 1) ++
-      pickRandomChars(upperLetters, 1) ++
-      pickRandomChars(numbers, 1) ++
-      pickRandomChars(specialChars, 1) ++
-      pickRandomChars(fullCharset, 12)
+    var password: String = pickRandomChars(fullCharset, passwordLength).mkString
+    // Keep generating passwords until we find one that has all of the required characters
+    // This is safer than picking from each subset and then shuffling
 
-    Random.shuffle(passwordChars).mkString
+    var isPasswordValid: Boolean =
+      password.exists(_.isLower) && password.exists(_.isUpper) && password.exists(_.isDigit) && password.exists(
+        specialChars.contains
+      )
+
+    while (isPasswordValid == false) {
+      password = pickRandomChars(fullCharset, passwordLength).mkString
+      isPasswordValid =
+        password.exists(_.isLower) && password.exists(_.isUpper) && password.exists(_.isDigit) && password.exists(
+          specialChars.contains
+        )
+
+    }
+    password
```

Comment:
hmm... can we do some testing about how many loops on average this takes? Dislike while loops a lot 😬

Reply:
Perhaps and approach like this, where we divide the number of chars into random segments based on the required categories so no looping is required: https://stackoverflow.com/questions/19743124/java-password-generator?

Reply:
Agreed, this is why I wrote the unit test using a length of 4, for that it takes on average about 10 loops. With a password length of 25 I never managed to even trigger the while loop in the first place

Reply:
thats fine then I guess, not worth spending too long as long as security team is happy...

Reply:
I will follow up with Sarah tomorrow. What you suggested is very close to what I initially implemented, but without the need for Java random.shuffle so it might be a good compromise, I really dislike while loops too!

Reply:
With a length of 25 I would expect it to not have to loop very often. This method is definitely more secure, and I would only want to consider less secure options if it had a performance impact.